internal/kms/kmip.go

```diff
@@ -0,0 +1,527 @@
+	protocolMinor = 4
+
+	// nonceSize is required to generate nonce for encrypting DEK.
+	nonceSize = 16
+
+	// kmipDefaultSecretsName is the default name of the Kubernetes Secret
+	// that contains the credentials to access the KMIP server. The name of
+	// the Secret can be configured by setting the `KMIP_SECRET_NAME`
+	// option.
+	//
+	// #nosec:G101, value not credential, just references token.
+	kmipDefaultSecretsName = "ceph-csi-kmip-credentials"
+
+	kmipEndpoint      = "KMIP_ENDPOINT"
+	kmipTLSServerName = "TLS_SERVER_NAME"
+	kmipReadTimeOut   = "READ_TIMEOUT"
+	kmipWriteTimeOut  = "WRITE_TIMEOUT"
+
+	// The following options are part of the Kubernetes Secrets.
+	//
+	// #nosec:G101, value not credential, just configuration keys.
+	kmipSecretNameKey    = "KMIP_SECRET_NAME"
+	kmipCACert           = "CA_CERT"
+	kmipCLientCert       = "CLIENT_CERT"
+	kmipClientKey        = "CLIENT_KEY"
+	kmipUniqueIdentifier = "UNIQUE_IDENTIFIER"
+)
+
+var _ = RegisterProvider(Provider{
+	UniqueID:    kmsTypeKMIP,
+	Initializer: initKMIPKMS,
+})
+
+type kmipKMS struct {
+	// basic options to get the secret
+	secretName string
+	namespace  string
+
+	// standard KMIP configuration options
+	endpoint         string
+	tlsConfig        *tls.Config
+	uniqueIdentifier string
+	readTimeout      uint8
+	writeTimeout     uint8
+}
+
+func initKMIPKMS(args ProviderInitArgs) (EncryptionKMS, error) {
+	kms := &kmipKMS{
+		namespace: args.Namespace,
+	}
+
+	// get secret name if set, else use default.
+	err := setConfigString(&kms.secretName, args.Config, kmipSecretNameKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = kmipDefaultSecretsName
+	}
+
+	err = setConfigString(&kms.endpoint, args.Config, kmipEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// optional
+	serverName := ""
+	err = setConfigString(&serverName, args.Config, kmipTLSServerName)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+
+	// optional
+	timeout := kmipDefaulfReadTimeout
+	err = setConfigInt(&timeout, args.Config, kmipReadTimeOut)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+	kms.readTimeout = uint8(timeout)
+
+	// optional
+	timeout = kmipDefaultWriteTimeout
+	err = setConfigInt(&timeout, args.Config, kmipWriteTimeOut)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+	kms.writeTimeout = uint8(timeout)
+
+	// read the Kubernetes Secret with CA cert, client cert, client key
+	// & key unique identifier.
+	secrets, err := kms.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secrets: %w", err)
+	}
+
+	caCert, found := secrets[kmipCACert]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCACert)
+	}
+
+	clientCert, found := secrets[kmipCLientCert]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, kmipCLientCert)
+	}
+
+	clientKey, found := secrets[kmipClientKey]
```

Comment:
the clientKey is used to decrypt the clientCert, what if the clientCert is not encrypted?

Reply:
> the clientKey is used to decrypt the clientCert, what if the clientCert is not encrypted?

This is the scenario that is currently tested/supported, 
we can maybe add support to such scenario as and when required after testing.

Reply:
sure, that's fine. I hope client certificates are encrypted in the common scenarios anyway.